src/Http/ReceivedAuthnRequestQueryString.php

```diff
@@ -0,0 +1,291 @@
+                    $queryParameter
+                ));
+            }
+
+            list($key, $value) = explode('=', $queryParameter, 2);
+
+            if (!in_array($key, self::$samlParameters)) {
+                continue;
+            }
+
+            if (array_key_exists($key, $parameters)) {
+                throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                    'Invalid ReceivedAuthnRequest query string ("%s"): parameter "%s" already present',
+                    $queryString,
+                    $key
+                ));
+            }
+
+            $parameters[$key] = $value;
+        }
+
+        if (!isset($parameters[self::PARAMETER_REQUEST])) {
+            throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                'Invalid ReceivedAuthnRequest query string ("%s"): parameter "%s" not found',
+                $queryString,
+                self::PARAMETER_REQUEST
+            ));
+        }
+
+        $parsedQueryString = new self($parameters[self::PARAMETER_REQUEST]);
+
+        if (isset($parameters[self::PARAMETER_RELAY_STATE])) {
+            $parsedQueryString->relayState = $parameters[self::PARAMETER_RELAY_STATE];
+        }
+        if (isset($parameters[self::PARAMETER_SIGNATURE])) {
+            if (base64_decode(urldecode($parameters[self::PARAMETER_SIGNATURE]), true) === false) {
+                throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                    'Invalid ReceivedAuthnRequest query string ("%s"): signature is not base64 encoded correctly',
+                    $queryString
+                ));
+            }
+
+            $parsedQueryString->signature = $parameters[self::PARAMETER_SIGNATURE];
+        }
+        if (isset($parameters[self::PARAMETER_SIGNATURE_ALGORITHM])) {
+            $parsedQueryString->signatureAlgorithm = $parameters[self::PARAMETER_SIGNATURE_ALGORITHM];
+        }
+
+        if ($parsedQueryString->signatureAlgorithm === null && $parsedQueryString->signature !== null) {
+            throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                'Invalid ReceivedAuthnRequest query string ("%s") contains a signature but not a signature algorithm',
+                $queryString
+            ));
+        }
+        if ($parsedQueryString->signature === null && $parsedQueryString->signatureAlgorithm !== null) {
+            throw new InvalidReceivedAuthnRequestQueryStringException(sprintf(
+                'Invalid ReceivedAuthnRequest query string ("%s") contains a signature algorithm but not a signature',
+                $queryString
+            ));
+        }
+
+        return $parsedQueryString;
+    }
+
+    /**
+     * @return bool
+     */
+    public function hasRelayState()
+    {
+        return $this->relayState !== null;
+    }
+
+    /**
+     * @return string
+     */
+    public function getSignedQueryString()
```

Comment:
Not sure, but I think this can be removed as we only really need the signableQueryString for signing and validating signatures.